pallets/stateful-storage/src/stateful_child_tree.rs

```diff
@@ -0,0 +1,395 @@
+use core::marker::PhantomData;
+
+use codec::{Decode, Encode};
+use common_primitives::msa::MessageSourceId;
+use frame_support::{
+	storage::{child, child::ChildInfo, ChildTriePrefixIterator},
+	Blake2_128, Blake2_256, Hashable, StorageHasher, Twox128, Twox256,
+};
+use sp_core::{ConstU8, Get};
+use sp_io::hashing::twox_64;
+use sp_std::{fmt::Debug, prelude::*};
+
+pub trait MultipartKeyStorageHasher: StorageHasher {
+	type HashSize: Get<u8>;
+	fn reverse(x: &[u8], num_key_parts: u8) -> &[u8] {
+		let hash_len: usize = (Self::HashSize::get() * num_key_parts).into();
+		if x.len() < hash_len {
+			log::error!("Invalid reverse: hash length too short");
+			return &[]
+		}
+		&x[hash_len..]
+	}
+}
+/// 128-bit Blake2 hash.
+impl MultipartKeyStorageHasher for Blake2_128 {
+	type HashSize = ConstU8<16>;
+}
+/// 256-bit Blake2 hash.
+impl MultipartKeyStorageHasher for Blake2_256 {
+	type HashSize = ConstU8<32>;
+}
+/// 128-bit XX hash.
+impl MultipartKeyStorageHasher for Twox128 {
+	type HashSize = ConstU8<16>;
+}
+/// 256-bit XX hash.
+impl MultipartKeyStorageHasher for Twox256 {
+	type HashSize = ConstU8<32>;
+}
+
+pub trait MultipartStorageKeyPart:
+	Clone + Debug + Default + Decode + Encode + Eq + PartialEq + Hashable
+{
+}
+impl<T> MultipartStorageKeyPart for T where
+	T: Clone + Debug + Default + Decode + Encode + Eq + PartialEq + Hashable
+{
+}
+
+pub trait MultipartKey<H: MultipartKeyStorageHasher>: MultipartStorageKeyPart {
+	type Arity: Get<u8>;
+
+	fn arity(&self) -> u8 {
+		Self::Arity::get()
+	}
+
+	fn hash(&self) -> Vec<u8>;
+	fn hash_prefix_only(&self) -> Vec<u8>;
+
+	fn decode(hash: &[u8]) -> Result<Self, codec::Error> {
+		let mut key_material = H::reverse(hash, Self::Arity::get());
```

Comment:
Did you mean to use the function defined above in used here?
```
fn arity(&self) -> u8 {
		Self::Arity::get()
	}
```

Reply:
nah, it wouldn't work with self. Removed the function